data-runtime/src/main/java/io/micronaut/data/runtime/mapper/sql/SqlResultEntityTypeMapper.java

```diff
@@ -443,6 +443,11 @@ private Object setChildrenAndTriggerPostLoad(Object instance, MappingContext<?>
                                     }
                                 }
                             }
+
+                            if (prop.getDataType() == DataType.JSON && jsonCodec != null) {
```

Comment:
I think this should be something like:
```
if (propertyType.isInstance(v)) {
                    args[i] = v;
                } else {
                    if (dataType == DataType.JSON && jsonCodec != null) {
                       args[i] = jsonCodec.decode(rpp.getArgument(), v.toString());
                    } else {
                        args[i] = resultReader.convertRequired(v, rpp.getArgument());
                    }
                }
```

Reply:
Thanks @dstepanov  I saw that last night when I was looking through broken tests. I will make the change for consistency now. I left it as is incase the JSON value needed to be converted, but this just seems like late night paranoia looking back on it.

I also noticed that the tests I added showed there was also an issue for R2DBC on H2 (no support for decoding JSON?). I have moved my new test 
cases into the existing JSON tests as that seemed to make more sense.

Reply:
I think H2 r2dbc driver doesn't support it yet